coremltools/converters/mil/mil/passes/defs/optimize_normalization.py

```diff
@@ -174,11 +175,23 @@ def _try_apply_transform(
         negative_axes = [a - rank if a >= 0 else a for a in axes]
         negative_axes.sort()
 
-        if len(gamma_var.val.shape) == len(axes) and len(beta_var.val.shape) == len(axes):
+        gamma_shape_len = len(gamma_var.val.shape) if gamma_var is not None else -1
+        beta_shape_len = len(beta_var.val.shape) if beta_var is not None else -1
+
+        if gamma_shape_len == len(axes) and beta_shape_len == len(axes):
             # axes for layer_norm must be [-1] or [-1, -2] or [-1, -2, -3] and so on
             if negative_axes == list(range(-len(negative_axes), 0)):
                 is_layernorm = True
 
+        if (
+            (gamma_var is None and beta_var is None)
+             or (gamma_shape_len == 1 and beta_shape_len == 1)
+        ):
+            if rank == 4 and negative_axes == [-3]:
+                is_layernorm = True
+                gamma_var = gamma_var.val if gamma_var else None
+                beta_var = beta_var.val if beta_var else None
+
```

Comment:
I think
```
if rank == 4 and negative_axes == [-3]:
   is_layernorm = (gamma_var is None and beta_var is None) or (gamma_rank == 1 and beta_rank == 1)
)
```
Is sufficient.
Note that, you don't need to materialize `gamma_var = gamma_var.val` at all (same for `beta_var`).
Since `mb.layer(..., gamma=gamma_var, beta=beta_var)` can accept type of `Var`

Reply:
That conditional looks good, using it now.

I tried not materializing gamma_var and beta_var, but my tests fail afterwards with:
`ValueError: Op 'mul_2' input gamma=_fuse_layernorm_gamma is not in scope of block27 before mul_2`

I'm not sure I fully understand the details here, but I'm constructing `gamma_var` and `beta_var` like this:
```python
gamma_var = mb.const(
    val=np.squeeze(gamma_var.val),
    name="_fuse_layernorm_gamma",
)
beta_var = mb.const(
    val=np.squeeze(beta_var.val) * gamma_var.val,
    name="_fuse_layernorm_beta"
)
```
Is something about that a problem / is there a better way?

Reply:
I think the reason is that we need to set the `before_op` when constructing `mb.const` properly ...
however I do see that we have the same mistake in the logic of other pattern for this pass ...
So your change looks good to me.

For detail if you are interested:
If you don't specify `before_op`, the new created `gamma_var, beta_var` are appended in the "last" of the graph, and they are invisible to the op `mul_2`.
When you create them, you need to set the `before_op` to the `mul` op, so they are topologically ahead of it and hence can be used as inputs.

Code reference: https://github.com/smpanaro/coremltools/blob/cf66920017c730dbeb2ec04dc88bbd9539d5c0c1/coremltools/converters/mil/mil/block.py#L318

Reply:
No actions need to be done.

Reply:
Ah, that is interesting, thanks for explaining. I'll leave it as-is for this PR, but seems like that would be a straightforward follow up PR?